# Data analysis of hypoxia in lake Buoy

This analysis project was done to determine all days and the corresponding time within 2018 where hypoxia was experienced in lake BOUY

In [1]:
# importing required libraries

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
# Loading the data set

buoy_data = pd.read_excel("buoy-2011-2020.xlsx", sheet_name = "Dissolved Oxygen")

## Analysis of the data

In [3]:
# Looking at the first five rows of the data to get an idea of what the data contains.

buoy_data.head()

Date/Time m/d/y H:M  DO Sat% 2m  DO Conc. (mg/L) 2m  DO Sat% 5m  \
0 2011-05-16 18:00:00        88.9                9.00        87.9   
1 2011-05-16 20:00:00        89.2                9.08        89.1   
2 2011-05-16 22:00:00        89.6                9.14        89.0   
3 2011-05-17 00:00:00        90.0                9.20        88.7   
4 2011-05-17 02:00:00        89.7                9.17        88.2   

   DO Conc. (mg/L) 5m  DO Sat% 8m  DO Conc. (mg/L) 8m  DO Sat% 11m  \
0                9.01        84.5                8.69         83.5   
1                9.08        86.3                8.84         83.9   
2                9.07        87.1                8.89         83.2   
3                9.06        86.9                8.88         83.6   
4                9.02        86.6                8.86         82.8   

   DO Conc. (mg/L) 11m  
0                 8.61  
1                 8.66  
2                 8.58  
3                 8.62  
4                 8.53

In [4]:
# Replacing NaT values in datetime column with NaN to aid in removing rows with only NULL values
buoy_data['Date/Time m/d/y H:M'] =  buoy_data['Date/Time m/d/y H:M'].astype(str)
buoy_data['Date/Time m/d/y H:M'] = buoy_data['Date/Time m/d/y H:M'].apply(lambda x : np.NaN if x=="NaT" else x) 

In [5]:
# removing rows that have only null values
buoy_data.dropna(how='all')

Date/Time m/d/y H:M  DO Sat% 2m  DO Conc. (mg/L) 2m  DO Sat% 5m  \
0       2011-05-16 18:00:00        88.9                9.00        87.9   
1       2011-05-16 20:00:00        89.2                9.08        89.1   
2       2011-05-16 22:00:00        89.6                9.14        89.0   
3       2011-05-17 00:00:00        90.0                9.20        88.7   
4       2011-05-17 02:00:00        89.7                9.17        88.2   
...                     ...         ...                 ...         ...   
140980  2020-08-25 09:15:00       108.4                8.95        99.5   
140981  2020-08-25 09:30:00       109.0                9.00       101.0   
140982  2020-08-25 09:45:00       108.4                8.96        94.6   
140983  2020-08-25 10:00:00       109.7                9.06       100.4   
140984  2020-08-25 10:15:00       110.2                9.11       100.8   

        DO Conc. (mg/L) 5m  DO Sat% 8m  DO Conc. (mg/L) 8m  DO Sat% 11m  \
0                     9.01        84.5                8.69         83.5   
1                     9.08        86.3                8.84         83.9   
2                     9.07        87.1                8.89         83.2   
3                     9.06        86.9                8.88         83.6   
4                     9.02        86.6                8.86         82.8   
...                    ...         ...                 ...          ...   
140980                8.26         7.3                0.73          7.0   
140981                8.38        14.3                1.38          6.8   
140982                7.88         9.7                0.98          6.7   
140983                8.33        16.8                1.59          6.8   
140984                8.36        14.6                1.39          6.8   

        DO Conc. (mg/L) 11m  
0                      8.61  
1                      8.66  
2                      8.58  
3                      8.62  
4                      8.53  
...                     ...  
140980                 0.74  
140981                 0.72  
140982                 0.71  
140983                 0.72  
140984                 0.72  

[140985 rows x 9 columns]

In [6]:
# creating Date and Time columns by Splitting the first column values into two

buoy_data['Date'] = pd.to_datetime(buoy_data['Date/Time m/d/y H:M']).dt.date
buoy_data['Time'] = pd.to_datetime(buoy_data['Date/Time m/d/y H:M']).dt.time
buoy_data.head()


Date/Time m/d/y H:M  DO Sat% 2m  DO Conc. (mg/L) 2m  DO Sat% 5m  \
0  2011-05-16 18:00:00        88.9                9.00        87.9   
1  2011-05-16 20:00:00        89.2                9.08        89.1   
2  2011-05-16 22:00:00        89.6                9.14        89.0   
3  2011-05-17 00:00:00        90.0                9.20        88.7   
4  2011-05-17 02:00:00        89.7                9.17        88.2   

   DO Conc. (mg/L) 5m  DO Sat% 8m  DO Conc. (mg/L) 8m  DO Sat% 11m  \
0                9.01        84.5                8.69         83.5   
1                9.08        86.3                8.84         83.9   
2                9.07        87.1                8.89         83.2   
3                9.06        86.9                8.88         83.6   
4                9.02        86.6                8.86         82.8   

   DO Conc. (mg/L) 11m        Date      Time  
0                 8.61  2011-05-16  18:00:00  
1                 8.66  2011-05-16  20:00:00  
2                 8.58  2011-05-16  22:00:00  
3                 8.62  2011-05-17  00:00:00  
4                 8.53  2011-05-17  02:00:00

In [7]:
buoy_data['Date'] = pd.to_datetime(buoy_data['Date'])

In [8]:
# getting the year value from the Date column and creating a new column year

buoy_data['year'] = buoy_data['Date'].dt.strftime('%Y')
buoy_data.head()

Date/Time m/d/y H:M  DO Sat% 2m  DO Conc. (mg/L) 2m  DO Sat% 5m  \
0  2011-05-16 18:00:00        88.9                9.00        87.9   
1  2011-05-16 20:00:00        89.2                9.08        89.1   
2  2011-05-16 22:00:00        89.6                9.14        89.0   
3  2011-05-17 00:00:00        90.0                9.20        88.7   
4  2011-05-17 02:00:00        89.7                9.17        88.2   

   DO Conc. (mg/L) 5m  DO Sat% 8m  DO Conc. (mg/L) 8m  DO Sat% 11m  \
0                9.01        84.5                8.69         83.5   
1                9.08        86.3                8.84         83.9   
2                9.07        87.1                8.89         83.2   
3                9.06        86.9                8.88         83.6   
4                9.02        86.6                8.86         82.8   

   DO Conc. (mg/L) 11m       Date      Time  year  
0                 8.61 2011-05-16  18:00:00  2011  
1                 8.66 2011-05-16  20:00:00  2011  
2                 8.58 2011-05-16  22:00:00  2011  
3                 8.62 2011-05-17  00:00:00  2011  
4                 8.53 2011-05-17  02:00:00  2011

#### selecting all rows for the year 2018

In [9]:
# filtering all the rows for the year 2018
buoy_data2 = buoy_data[buoy_data["year"] == '2018']
buoy_data2

Date/Time m/d/y H:M  DO Sat% 2m  DO Conc. (mg/L) 2m  DO Sat% 5m  \
115613  2018-04-26 15:00:00       100.4               11.58        97.2   
115614  2018-04-26 15:15:00        99.5               11.56        97.1   
115615  2018-04-26 15:30:00       100.6               11.62        97.6   
115616  2018-04-26 15:45:00       100.0               11.60        97.2   
115617  2018-04-26 16:00:00        98.5               11.52        97.6   
...                     ...         ...                 ...         ...   
128025  2018-10-30 08:00:00        87.3                9.98        88.3   
128026  2018-10-30 08:15:00        87.2                9.97        88.2   
128027  2018-10-30 08:30:00        87.1                9.96        88.3   
128028  2018-10-30 08:45:00        87.2                9.96        88.1   
128029  2018-10-30 09:00:00        87.3                9.97        88.3   

        DO Conc. (mg/L) 5m  DO Sat% 8m  DO Conc. (mg/L) 8m  DO Sat% 11m  \
115613               11.49        96.5               11.55         92.7   
115614               11.49        97.1               11.61         92.5   
115615               11.56        96.6               11.58         92.4   
115616               11.51        96.6               11.59         92.3   
115617               11.57        96.4               11.57         92.4   
...                    ...         ...                 ...          ...   
128025               10.08        87.1               10.00          NaN   
128026               10.07        87.5               10.05          NaN   
128027               10.08        87.5               10.04          NaN   
128028               10.06        87.3               10.02          NaN   
128029               10.09        87.8               10.08          NaN   

        DO Conc. (mg/L) 11m       Date      Time  year  
115613                11.62 2018-04-26  15:00:00  2018  
115614                11.61 2018-04-26  15:15:00  2018  
115615                11.60 2018-04-26  15:30:00  2018  
115616                11.60 2018-04-26  15:45:00  2018  
115617                11.61 2018-04-26  16:00:00  2018  
...                     ...        ...       ...   ...  
128025                  NaN 2018-10-30  08:00:00  2018  
128026                  NaN 2018-10-30  08:15:00  2018  
128027                  NaN 2018-10-30  08:30:00  2018  
128028                  NaN 2018-10-30  08:45:00  2018  
128029                  NaN 2018-10-30  09:00:00  2018  

[12417 rows x 12 columns]

In [10]:
# selecting my desired columns
buoy_data3 = buoy_data2[['Date/Time m/d/y H:M', 'DO Conc. (mg/L) 2m', 'DO Conc. (mg/L) 5m', 'DO Conc. (mg/L) 8m', 'DO Conc. (mg/L) 11m']]
buoy_data3

Date/Time m/d/y H:M  DO Conc. (mg/L) 2m  DO Conc. (mg/L) 5m  \
115613  2018-04-26 15:00:00               11.58               11.49   
115614  2018-04-26 15:15:00               11.56               11.49   
115615  2018-04-26 15:30:00               11.62               11.56   
115616  2018-04-26 15:45:00               11.60               11.51   
115617  2018-04-26 16:00:00               11.52               11.57   
...                     ...                 ...                 ...   
128025  2018-10-30 08:00:00                9.98               10.08   
128026  2018-10-30 08:15:00                9.97               10.07   
128027  2018-10-30 08:30:00                9.96               10.08   
128028  2018-10-30 08:45:00                9.96               10.06   
128029  2018-10-30 09:00:00                9.97               10.09   

        DO Conc. (mg/L) 8m  DO Conc. (mg/L) 11m  
115613               11.55                11.62  
115614               11.61                11.61  
115615               11.58                11.60  
115616               11.59                11.60  
115617               11.57                11.61  
...                    ...                  ...  
128025               10.00                  NaN  
128026               10.05                  NaN  
128027               10.04                  NaN  
128028               10.02                  NaN  
128029               10.08                  NaN  

[12417 rows x 5 columns]

In [11]:
# converting the shown column to int to enable selection of all rows with any value less than 2 
buoy_data3['Date/Time m/d/y H:M'] =  pd.to_datetime(buoy_data3['Date/Time m/d/y H:M']).astype(np.int64)

C:\Users\MAURINE\AppData\Local\Temp\ipykernel_1840\3386071403.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  buoy_data3['Date/Time m/d/y H:M'] =  pd.to_datetime(buoy_data3['Date/Time m/d/y H:M']).astype(np.int64)


In [12]:
# selecting all rows with any values less than 2 from any of the columns
buoy_data10 = buoy_data3[(buoy_data3<2).any(axis=1)]
buoy_data10

Date/Time m/d/y H:M  DO Conc. (mg/L) 2m  DO Conc. (mg/L) 5m  \
119424  1531756800000000000               10.38                6.65   
119425  1531757700000000000               10.81                6.69   
119426  1531758600000000000               10.74                6.64   
119427  1531759500000000000               10.68                6.52   
119428  1531761300000000000               10.47                6.17   
...                     ...                 ...                 ...   
124619  1537346700000000000               10.04                5.76   
124620  1537347600000000000                9.98                8.88   
124624  1537351200000000000                9.39                7.94   
124625  1537352100000000000                9.32                6.61   
124626  1537353000000000000                9.20                8.28   

        DO Conc. (mg/L) 8m  DO Conc. (mg/L) 11m  
119424                2.29                 0.65  
119425                1.82                 0.98  
119426                1.65                 1.22  
119427                1.74                 1.23  
119428                1.50                 1.28  
...                    ...                  ...  
124619                1.94                  NaN  
124620                1.94                  NaN  
124624                1.94                  NaN  
124625                1.87                  NaN  
124626                1.97                  NaN  

[3986 rows x 5 columns]

### Final data set showing all days and the corresponding  time within 2018 where hypoxia was esperienced i.e dissolved oxygen levels below 2mg/L
The table shows a total of 3986 entries 

In [13]:
# converting the column shown below back to date time.
buoy_data10['Date/Time m/d/y H:M'] =  pd.to_datetime(buoy_data10['Date/Time m/d/y H:M'])
buoy_data10

C:\Users\MAURINE\AppData\Local\Temp\ipykernel_1840\2018388295.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  buoy_data10['Date/Time m/d/y H:M'] =  pd.to_datetime(buoy_data10['Date/Time m/d/y H:M'])


Date/Time m/d/y H:M  DO Conc. (mg/L) 2m  DO Conc. (mg/L) 5m  \
119424 2018-07-16 16:00:00               10.38                6.65   
119425 2018-07-16 16:15:00               10.81                6.69   
119426 2018-07-16 16:30:00               10.74                6.64   
119427 2018-07-16 16:45:00               10.68                6.52   
119428 2018-07-16 17:15:00               10.47                6.17   
...                    ...                 ...                 ...   
124619 2018-09-19 08:45:00               10.04                5.76   
124620 2018-09-19 09:00:00                9.98                8.88   
124624 2018-09-19 10:00:00                9.39                7.94   
124625 2018-09-19 10:15:00                9.32                6.61   
124626 2018-09-19 10:30:00                9.20                8.28   

        DO Conc. (mg/L) 8m  DO Conc. (mg/L) 11m  
119424                2.29                 0.65  
119425                1.82                 0.98  
119426                1.65                 1.22  
119427                1.74                 1.23  
119428                1.50                 1.28  
...                    ...                  ...  
124619                1.94                  NaN  
124620                1.94                  NaN  
124624                1.94                  NaN  
124625                1.87                  NaN  
124626                1.97                  NaN  

[3986 rows x 5 columns]